qiskit_experiments/framework/experiment_data.py

```diff
@@ -795,6 +797,116 @@ def add_data(
                     self._add_result_data(datum)
                 else:
                     raise TypeError(f"Invalid data type {type(datum)}.")
+    
+    def _add_data(
```

Comment:
You need to update this line instead. Having a dedicated method for marginalization is also okey.
https://github.com/Qiskit-Extensions/qiskit-experiments/blob/f3a26684196b8acb2b0ca942cb7b262ce33c9c3c/qiskit_experiments/framework/experiment_data.py#L793

The returned Qiskit `Result` object is formatted into a canonical dictionary format, which contains experiment data and metadata (the metadata is created by the experiment class). When you run a composite experiment, your experiment object gains a special metadata field
https://github.com/Qiskit-Extensions/qiskit-experiments/blob/f3a26684196b8acb2b0ca942cb7b262ce33c9c3c/qiskit_experiments/framework/composite/composite_experiment.py#L193-L201 
The structure looks ugly when you nest composite experiment -- this indicates you need to recursively check the field, e.g. `datum["metadata"]["component_metadata"][0]["metadata"]["component_metadata"][0]...` -- I want to refactor this at some point, but not for now.

You would check the `datum["metadata"]`. When these fields appear, you can instantiate component experiment data and [set them to the current instance](https://github.com/Qiskit-Extensions/qiskit-experiments/blob/f3a26684196b8acb2b0ca942cb7b262ce33c9c3c/qiskit_experiments/framework/experiment_data.py#L2340-L2345). Then, you check next `datum`, and crate new component data when the corresponding container is missing (for example, if you run batch experiment of two parallel experiments, first half of data contains result for half of qubits, and latter half contains the rest of qubits). This is just a clue for implementation, so you can try your own idea if you have in mind :)